torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1550,10 +1550,64 @@ def _verify(expr, potential_expr):
 
         # 2. Every guard must evaluate to True (but remember many guards
         #    like s0 == s1*2 because trivial due to simplification)
-        for g, tb in self.guards:
-            if self._maybe_evaluate_static(g) is not None:
+        def simplify_until_no_change(expr: sympy.Expr) -> sympy.Expr:
+            new_expr = self.simplify(expr)
+            while new_expr != expr:
+                expr = new_expr
+                new_expr = self.simplify(expr)
+            return expr
+
+        simplified_guards = [(simplify_until_no_change(g), tb) for g, tb in self.guards]
+        var_to_guards = dict()
+        var_to_range = self.var_to_range.copy()
+        unhandled_guards = []
+
+        for g, tb in simplified_guards:
+            if self._maybe_evaluate_static_impl(g, var_to_range=var_to_range) is not None:
+                continue
+
+            if not isinstance(g, sympy.Rel):
+                unhandled_guards.append((g, tb))
+                continue
+
+            symbol = g.lhs
+
+            if not isinstance(g.lhs, sympy.Symbol):
+                unhandled_guards.append((g, tb))
                 continue
-            g = self.simplify(g)
+
+            assert symbol in self.var_to_range
+
+            # Consider actually issuing the guard if it is used for refining
+            # the value range of a symbol. Otherwise, it is discarded.
```

Comment:
Refining the ranges at the end seems worse. You want to update the ranges AS SOON as you know you have a refined range! Do it earlier, it will save work!